src/wormhole/Governance.sol

```diff
@@ -0,0 +1,92 @@
+// SPDX-License-Identifier: Apache 2
+pragma solidity >=0.6.12 <0.9.0;
+
+import "wormhole-solidity-sdk/interfaces/IWormhole.sol";
+
+contract Governance {
+    IWormhole immutable wormhole;
+
+    error InvalidGovernanceChainId(uint16 chainId);
+    error InvalidGovernanceContract(bytes32 contractAddress);
+    error GovernanceActionAlreadyConsumed(bytes32 digest);
+
+    error NotRecipientChain(uint16 chainId);
+    error NotRecipientContract(address contractAddress);
+
+    bytes32 constant CONSUMED_GOVERNANCE_ACTIONS_SLOT =
+        keccak256("governance.consumedGovernanceActions");
+
+    function _getConsumedGovernanceActionsStorage()
+        private
+        pure
+        returns (mapping(bytes32 => bool) storage $)
+    {
+        uint256 slot = uint256(CONSUMED_GOVERNANCE_ACTIONS_SLOT);
+        assembly ("memory-safe") {
+            $.slot := slot
+        }
+    }
+
+    struct GovernanceMessage {
+        uint16 chainId;
+        address governanceContract;
+        address governedContract;
+        bytes callData;
+    }
+
+    constructor(address _wormhole) {
+        wormhole = IWormhole(_wormhole);
+    }
+
+    function performGovernance(bytes calldata vaa) external {
+        IWormhole.VM memory verified = _verifyGovernanceVAA(vaa);
+        GovernanceMessage memory message = abi.decode(verified.payload, (GovernanceMessage));
+
+        if (message.chainId != wormhole.chainId()) {
+            revert NotRecipientChain(message.chainId);
+        }
+        if (message.governanceContract != address(this)) {
+            revert NotRecipientContract(message.governanceContract);
+        }
```

Comment:
Can we get rid of this check and instead add a "module" hash to the `GovernanceMessage` struct. And then verify that `module == keccak256("NativeTokenTransfers")` or something like that?

My thinking is we could then get the guardians to sign a single message with `chainId` of 0 to perform an action that is relevant for every chain (e.g. calling `pause`). We could restrict this to specific actions by checking the first 4 bytes of calldata once we've finalised all the allowed actions of the Manager owner.

Maybe this isn't too important since we won't have loads of deployments, lmk what you think?

Reply:
That's a good point, there might be actions that make sense on all chains. My original reason for putting the address in was that if this contract ever needs to be migrated to another one that works similarly (perhaps a bug fix), then with the address in, we wouldn't need to worry about replays. Maybe that's just too paranoid?

Reply:
Hmmm good point, that's the downside of not making this contract upgradable. Let's leave it as is then, I'd rather have there be a little more effort for governors than have the potential of replaying these actions.

Or we make this upgradable to solve both? I originally thought this contract would look a bit different but since it's making arbitrary external calls anyway I don't think making this upgradable would introduce any new attack vectors?